File: Algorithms/C++/33. Search in Rotated Sorted Array.cpp
// Runtime: 8 ms, faster than 31.30% of C++ online submissions for Search in Rotated Sorted Array.
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0, end = nums.size();
        
        while (start != end) {  // 下标可能在左右侧进行跳转，所以不能用<>进行判断
            const int mid = start + (end-start) / 2; // 为啥这里用const？
            if (nums[mid] == target) {  // 最简单的一种情况
                return mid;
            } 
            if (nums[start] <= nums[mid]) { // 从start到mid处于有序状态，考虑了重复的元素？
                if (nums[start] <= target && target < nums[mid]) // 如果target在此之间
                    end = mid;
                else 
                    start = mid + 1;
            } else {    // 从mid到end处于有序状态
                if (nums[mid] < target && target <= nums[end - 1])
                    start = mid + 1;
                else 
                    end = mid;
            }
        }
        
        return -1;
    }
};

File: Algorithms/Java/33. Search in Rotated Sorted Array.java
class Solution {
    public int search(int[] nums, int target) {
        if (nums.length == 0) return -1;
        if (nums.length == 1) return (nums[0] == target) ? 0 : -1;
        
        int left = 0, right = nums.length - 1;
        
        while (left != right) { // the left + 1 = right
            int mid = left + (right-left) / 2;
            if (nums[mid] == target) return mid;
            if (nums[left] <= nums[mid]) {     // nums[start..mid] sorted
                if (target >= nums[left] && target < nums[mid]) {   // if target in the sorted part
                    right = mid;
                } else {
                    left = mid + 1;
                }
            } else {  // nums[mid..end] sorted
                if (target > nums[mid] && target <= nums[right]) {
                    left = mid;
                } else {
                    right = mid - 1;
                }
            }
        } // END WHILE
        
        if (target == nums[left]) return left;
        if (target == nums[right]) return right;
        
        return -1;
    }
}